docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1911 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #57.            add_cand_i = include[i] 
</pre><p>I don't think we need this line, right? What I mean is... can't we just use <code>include[i]</code> everywhere? So, we'd replace <code>add_cand_i</code> with:</p><pre class="ql-syntax" spellcheck="false">if len(mask):
    include[i] = False
</pre><p>and then:</p><pre class="ql-syntax" spellcheck="false">if add_cand_i:
    is_cands[i] = True
</pre><p>turns into a single line:</p><pre class="ql-syntax" spellcheck="false">is_cands[i] = include[i]
</pre><p> </p><p>I'm really starting to wonder if <code>is_cands</code> is even needed?? It seems like it is redundant with <code>include</code> or there is some sort of shared information? If I squint my eyes, it almost feels like all that we're doing is updating <code>include</code> . Maybe not?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Reply:
> Line #57.            add_cand_i = include[i] 
> I don't think we need this line, right? What I mean is... can't we just use `include[i]` everywhere? 
```
if len(mask):
    include[i] = False
```
Doesn't  this change the include in the outer wrapper? If yes, then, is it better to do `include = include.copy()` in the favor of removing `if` in `if add_cand_i: ...`

---

> I'm really starting to wonder if `is_cands` is even needed??

I think we can do it. Just to let you know:  `include` basically contains the indices that potentially can be discord. `is_cands`, however, depends on the `min_dist`. So if the algorithm cannot find any candidate with LARGE `min_dist`, it needs to lower `min_dist` and scan the indices of `include` again.

So, we can do `include = include.copy()` so that we do not change the `include` of outer wrapper. We update `include` inside the `_select_candidates` and return it. But, we have to use a new name for the output when it is being called in the outer wrapper as we don't want to change `include`. So, I think the current version is cleaner as it deliver the message by the name `is_cands`. What do you think?

Reply:
@seanlaw 
Does this work ? (this is basically what you suggested)

```
# in `_select_candidates`

def _select_candidates(...):
    include = include.copy()
    ...
    # your suggestion: removing `add_cand_i`
```

And, keep the variable `is_cands`.

Reply:
As I don't want to change `include`, I tried:

```
is_cands = include.copy()
```

and then just do,

```
if len(mask):
    is_cands[i] = False
```

And, we do not need the line `is_cands[i] = include[i]` anymore!

It is working (I tried it with different case studies).

Reply:
<p>What about this?</p><pre class="ql-syntax" spellcheck="false">if add_cand_i:
    is_cands[i] = True
</pre>

Reply:
<blockquote>So, we can do&nbsp;<code>include = include.copy()</code>&nbsp;so that we do not change the&nbsp;<code>include</code>&nbsp;of outer wrapper. We update&nbsp;<code>include</code>&nbsp;inside the&nbsp;<code>_select_candidates</code>&nbsp;and return it. But, we have to use a new name for the output when it is being called in the outer wrapper as we don't want to change&nbsp;<code>include</code>. So, I think the current version is cleaner as it deliver the message by the name&nbsp;<code>is_cands</code>. What do you think?</blockquote><p>I see what you mean about not wanting to change <code>include</code> . Then, to be clear, let's not copy from <code>include</code>  and only read from <code>include</code>  :</p><p> </p><pre class="ql-syntax" spellcheck="false">n&nbsp;= T.shape[0]&nbsp;
&nbsp; &nbsp; k = n - m + 1 # number of subsequences
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; T_subseq_isfinite = core.rolling_isfinite(T, m)
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; is_cands = np.zeros(k, dtype=bool)
&nbsp; &nbsp; excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; for i in np.flatnonzero(T_subseq_isfinite):
&nbsp; &nbsp; &nbsp; &nbsp; add_cand_i = include[i]&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; if i &gt; excl_zone:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; left_non_trivial_cand_idx = np.flatnonzero(is_cands[:i-excl_zone])
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; cand_idx_chunks = _get_chunks_ranges(left_non_trivial_cand_idx) # get left_non_trivial cand indices in chunks
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; for (start, stop) in cand_idx_chunks:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Q = T[i:i+m]
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; QT = core.sliding_dot_product(Q, T[start: (stop+m-1)]) #stop + m - 1 is correct for converting...&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; #stop from "subsequence index" space to "time series index" space...&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; #Do not replace with the incorrect stop - m + 1 in this case!
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; D = core._mass(Q, T[start: (stop+m-1)], QT, M_T[i], Σ_T[i], M_T[start:stop], Σ_T[start:stop])
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mask = np.flatnonzero(D &lt; min_dist)&nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; is_cands[start:stop][mask] = False


&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; if len(mask):
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; add_cand_i = False
&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; if add_cand_i:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; is_cands[i] = True
</pre>

Reply:
> let's not copy from `include` and only read from `include` :

Just out of curiosity, will there be a problem if we do `is_cands = include.copy()`?  Because, I used the following code below:

```
    is_cands = include.copy()

    n = T.shape[0] 
    k = n - m + 1 # number of subsequences
    
    T_subseq_isfinite = core.rolling_isfinite(T, m)
    
    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
    
    for i in np.flatnonzero(T_subseq_isfinite):        
        if i > excl_zone:
            left_non_trivial_cand_idx = np.flatnonzero(is_cands[:i-excl_zone])
            cand_idx_chunks = _get_chunks_ranges(left_non_trivial_cand_idx) # get left_non_trivial cand indices in chunks
            for (start, stop) in cand_idx_chunks:
                Q = T[i:i+m]
                QT = core.sliding_dot_product(Q, T[start: (stop+m-1)]) #stop + m - 1 is correct for converting... 
                #stop from "subsequence index" space to "time series index" space... 
                #Do not replace with the incorrect stop - m + 1 in this case!
                D = core._mass(Q, T[start: (stop+m-1)], QT, M_T[i], Σ_T[i], M_T[start:stop], Σ_T[start:stop])
                
                mask = np.flatnonzero(D < min_dist)   
                is_cands[start:stop][mask] = False

                if len(mask):
                    is_cands[i] = False
        
       
    return is_cands
```

So, do you think `if add_cand_i and include[i]:` make it more readable?

Reply:
> I think I prefer this more because the intention of how you are "using" `include` is much clearer and that it should be completely independent of `add_cand_i` .

I see it now. So, I assume you are using one more `if` to get the code more readable. will be revised in next set of commits. Thanks for the suggestion!

Reply:
> Just out of curiosity, will there be a problem if we do is_cands = include.copy()? 

I don't see anything obviously wrong but it may be hard to spot. You are better off testing with a large variety of random inputs just to at least convince yourself that your code has seen enough different examples. I don't have any problem with your proposal but please test.

Reply:
Both works. I tested it several times on different randomly-generated time series (I took a look at the whole process and not just final results. They give EXACTLY same output). In terms of computation time, they are almost the same!

So, I looked through both approaches several times, and I think your idea ` if add_cand_i and include[i]:` is easier in terms of readability. I would go with that. However, if you think otherwise, please let me know.